### PR TITLE
enhance: Support auto index for scalar index

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #29309 
reopen pr : #29310

### CODE REVIEW

File: internal/proxy/task_index.go [hunk 1]
```diff
@@ -144,36 +144,29 @@ func (cit *createIndexTask) parseIndexParams() error {
...
+		specifyIndexType, exist := indexParamsMap[common.IndexTypeKey]
+		if Params.AutoIndexConfig.ScalarAutoIndexEnable.GetAsBool() || specifyIndexType == AutoIndexName || !exist {
+			if typeutil.IsArithmetic(cit.fieldSchema.DataType) {
+				indexParamsMap[common.IndexTypeKey] = Params.AutoIndexConfig.ScalarNumericIndexType.GetValue()
+			} else if typeutil.IsStringType(cit.fieldSchema.DataType) {
+				indexParamsMap[common.IndexTypeKey] = Params.AutoIndexConfig.ScalarVarcharIndexType.GetValue()
+			} else if typeutil.IsBoolType(cit.fieldSchema.DataType) {
+				indexParamsMap[common.IndexTypeKey] = Params.AutoIndexConfig.ScalarBoolIndexType.GetValue()
+			} else {
+				return merr.WrapErrParameterInvalid("supported field",
```
Comment (longjiquan) [inline internal/proxy/task_index.go:167]: This error message is not clear, will mislead users to think the data type can't be indexed. Maybe we want to let users to know we don't support auto index for this data type?

File: internal/proxy/util.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ const (
 	DefaultArithmeticIndexType = "STL_SORT"
 
 	// DefaultStringIndexType name of default index type for varChar/string field
-	DefaultStringIndexType = "Trie"
```
Comment (longjiquan) [inline internal/proxy/util.go:71]: Do not edit this for better compatability.

File: pkg/util/indexparamcheck/index_type.go [hunk 3]
```diff
@@ -32,6 +32,13 @@ const (
 	IndexDISKANN         IndexType = "DISKANN"
 	IndexSparseInverted  IndexType = "SPARSE_INVERTED_INDEX"
 	IndexSparseWand      IndexType = "SPARSE_WAND"
+	IndexINVERTED        IndexType = "INVERTED"
+
+	IndexSTLSORT IndexType = "STL_SORT"
+	IndexTRIE    IndexType = "TRIE"
```
Comment (longjiquan) [inline pkg/util/indexparamcheck/index_type.go:38]: Remove this.

File: pkg/util/paramtable/autoindex_param.go [hunk 4]
```diff
@@ -38,6 +38,12 @@ type autoIndexConfig struct {
 	AutoIndexTypeName     ParamItem  `refreshable:"true"`
 	AutoIndexSearchConfig ParamItem  `refreshable:"true"`
 	AutoIndexTuningConfig ParamGroup `refreshable:"true"`
+
+	ScalarAutoIndexEnable  ParamItem `refreshable:"true"`
+	ScalarAutoIndexParams  ParamItem `refreshable:"true"`
+	ScalarNumericIndexType ParamItem `refreshable:"true"`
+	ScalarVarcharIndexType ParamItem `refreshable:"true"`
+	ScalarBoolIndexType    ParamItem `refreshable:"true"`
```
Comment (longjiquan) [inline pkg/util/paramtable/autoindex_param.go:46]: Quick question: what if these index types are invalid?

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: pkg/util/indexparamcheck/index_type.go (hunk 3 above)
Comment (xiaocai2333) [reply pkg/util/indexparamcheck/index_type.go:38]: @czs007

File: internal/proxy/task_index.go [hunk 5]
```diff
@@ -144,36 +144,29 @@ func (cit *createIndexTask) parseIndexParams() error {
...
-		} else if typeutil.IsBoolType(cit.fieldSchema.DataType) {
-			if !exist {
-				return merr.WrapErrParameterInvalidMsg("no index type specified")
-			}
-			if specifyIndexType != InvertedIndexType {
-				return merr.WrapErrParameterInvalidMsg("index type (%s) not supported for boolean, supported: %s",
-					specifyIndexType, InvertedIndexType)
+	if !isVecIndex {
+		specifyIndexType, exist := indexParamsMap[common.IndexTypeKey]
+		if Params.AutoIndexConfig.ScalarAutoIndexEnable.GetAsBool() || specifyIndexType == AutoIndexName || !exist {
```
[the remark this answers is not in the document]
Comment (xiaocai2333) [reply internal/proxy/task_index.go:159]: yes, the behavior is consistent with AUTOINDEX of the vector.

File: internal/proxy/util.go (hunk 2 above)
Comment (xiaocai2333) [reply internal/proxy/util.go:71]: This was decided by @czs007 , so all our index types are in uppercase.

File: pkg/util/paramtable/autoindex_param.go (hunk 4 above)
Comment (xiaocai2333) [reply pkg/util/paramtable/autoindex_param.go:46]: will return an error.

Comment (xiaocai2333) [reply pkg/util/paramtable/autoindex_param.go:46]: like
```
<MilvusException: (code=1100, message=invalid parameter[expected=valid index][actual=invalid index type: trie])>
```

**Comment by czs007** [comment]
@xiaocai2333 need cherry-pick to branch 2.4

**Pull request merged**